### ISSUE TITLE
Improve dining location sorting

### ISSUE THREAD

**Issue by c3bryant** [UCSD/campus-mobile#358]
Dining location results were originally sorted alphanumerically. Over time, users showed a stronger desire to have the results sorted by distance and we moved to that. Currently, neither sort method by itself is ideal, and neither method is working as expected in prod or dev. 

This task is to:
1. Add two Dining card menu options, above `Hide Card`:
- `Sort by Closest`
- `Sort A-Z`
![screen shot 2018-11-27 at 12 38 59 pm](https://user-images.githubusercontent.com/15112294/49110715-13940b00-f243-11e8-945b-f824a32dcf13.png)

2. Implement `Sort By` for Dining -> View All:
![diningcardsort](https://user-images.githubusercontent.com/15112294/49110170-96b46180-f241-11e8-979c-96afbf86b9a2.png)

3. Save the user's dining sort preference locally always, and remotely if logged in

Notes:
On Dining -> View All, the Sort By options should scroll with the view

**Comment by g9singh** [comment]
completed in: 78631322a310d2f57f747981695b3a83d48a7d2d

**Comment by c3bryant** [comment]
Updates made and verified in 08c29c6e on branch `feature/dining-location-sorting`

**Comment by c3bryant** [comment]
Merged into `refresh-2`

**Comment by c3bryant** [comment]
Further QA testing failed with one issue:
* Sort by state doesn't always match highlighted Sort By selection

![dining-sort-inconsistent](https://user-images.githubusercontent.com/15112294/54956577-88373780-4f0d-11e9-9fe3-3d9c386e387b.png)

Testing revealed that _sortDining in diningSaga was being passed null and entering the default switch statement

Next steps:
* Change the sort by initial state to A-Z, as this is available to all users, vs location services ('Closest') which must be opted into
* Consider hiding the 'Sort By' bar if the user does not have location services enabled
* Consider greying out the 'Closest' sort by option if the user does not have location services enabled
Adding @woznaut

**Comment by c3bryant** [comment]
Removed from branch `refresh-2`

**Comment by g9singh** [comment]
Update to requirements: 
Default to sort by Closest if user has location enabled

Default to A-Z if user has not enabled location and grey out Closest button
Also if grey Closest button is clicked the user should be prompted for location access again

**Comment by sssaiya** [comment]
Requirements after discussion:
- Grey out / Don't grey out closest button when location is off (Show both options to team).
- Show prompt asking for location to be turned on when user clicks on sort by closest.
- Default to sort by a-z if loc not enabled.
- Show '-' instead of distance when location disabled.

**Comment by sssaiya** [comment]
Features implemented:
- Show prompt asking for location services to be turned on when user clicks on closest with their location currently turned on.
- doesn't allow user to select closest when location not on.
- Shows '-' instead of distances when loc is disabled.
- Defaults to a-z if location is disabled.

Remaining:
Grey out / Don't grey out closest button when location is off (Show both options to team).
distance update to '-' takes dining TTL time, should maybe be instant.

**Comment by sssaiya** [comment]
Updated in commit c66a981:
- Add locationListener in a LocationContainer.
- Add location status to state.
- DiningDistances update correctly depending on location being on or off.

remaining:
- Bug when on closest sort and location turned off, component doesn't re-render.

**Comment by sssaiya** [comment]
Completed in [67ba4c8](https://github.com/UCSD/campus-mobile/commit/67ba4c8b5e39943f7e7d6fd0348bb74b263352be)

**Comment by g9singh** [comment]
Needed to install a new library that allowed for an event to trigger when user disabled location. Also required some rework after library was installed. 
Completed on branch feature/dining-location-sorting
178e6a80bf0bab2b0127b613a45bf644dcfae51f

**Comment by g9singh** [comment]
changes made: 
Reword alert message when user has location services turned off
Fixed issue where android devices were not able to switch to 'Closest' without turning location services on/off
small cosmetic changes in dining sort bar

commit: bfa23abb82119dec7fed0359a8c3f5dee5bea134

**Comment by c3bryant** [comment]
Test results:

* PASS on iOS with no issues
* FAIL on Android with two issues (build 625):
1) Initial load never shows dining distances on Dining card (shows '-')
2) Unable to activate `Closest` sort button on Dining View All (greyed out regardless of settings)

Notes:
6.6 for Android (build 581) in production does not experience either of the two issues noted above

**Comment by vishant101** [comment]
Fixed issues with android on ae8c164445863936748b77042e2240b77444edf2

@c3bryant @g9singh  please confirm changes work and do not break functionality for iOS

**Comment by g9singh** [comment]
Upon further testing the issue persisted. 

Setup: A fresh install on an android device with location enabled (os level) and access given to the app upon launch

1) Dining distances on Dining card (shows '-')
2) Unable to activate Closest sort button on Dining View All (greyed out)

Other notes:
1) If location is turned of and then back on the distances start to appear
2) After about a minute of just waiting on the Dining card the distances turn into dashes. This makes it seem as if the location is disabled.
3) After another few seconds the distances reappear. The ordering of the list remains unchanged throughout the entire process.

**Comment by vishant101** [comment]
@g9singh notes 2/3 maybe caused by the location data updating. Checking the value of the enabled for location shows that it doesn't change (the dashes may be appearing as a result of the distances not being available). The app update which makes the app data driven may ultimately fix those issue.

**Comment by c3bryant** [comment]
This issue is being closed due to platform migration and may be revisited at a later date.
